benchmarks/serving/backend_request_func.py

```diff
@@ -0,0 +1,216 @@
+            "temperature": 0.0,
+            "top_p": 1.0,
+            "max_tokens": output_len,
+            "stream": False,
+        }
+        output = {}
+        output["prompt_len"] = prompt_len
+
+        st = time.perf_counter()
+        async with session.post(url=api_url, json=payload) as resp:
+            if resp.status == 200:
+                parsed_resp = await resp.json()
+                latency = time.perf_counter() - st
+                output["generated_text"] = parsed_resp["text_output"]
+                output["success"] = True
+                output["latency"] = latency
+            else:
+                output["generated_text"] = ""
+                output["success"] = False
+
+        return output
+
+
+async def async_request_deepspeed_mii(
+    prompt: str,
+    api_url: str,
+    prompt_len: int,
+    output_len: int,
+    best_of: int,
+    use_beam_search: bool,
+    **kwargs,
+):
+    timeout = aiohttp.ClientTimeout(total=6 * 60 * 60)
+
+    async with aiohttp.ClientSession(timeout=timeout) as session:
+        assert best_of == 1
+        assert not use_beam_search
+
+        payload = {
+            "prompt": prompt,
+            "max_new_tokens": output_len,
+            "ignore_eos": True,
+            "do_sample": True,
+            "temperature": 0.01,  # deepspeed-mii does not accept 0.0 temperature.
+            "top_p": 1.0,
+        }
+        output = {}
+        output["prompt_len"] = prompt_len
+
+        st = time.perf_counter()
+        async with session.post(url=api_url, json=payload) as resp:
+            if resp.status == 200:
+                parsed_resp = await resp.json()
+                latency = time.perf_counter() - st
+                output["generated_text"] = parsed_resp[0]["generated_text"]
+                output["success"] = True
+                output["latency"] = latency
+            else:
+                output["generated_text"] = ""
+                output["success"] = False
+
+        return output
+
+
+async def async_request_openai_completions(
+    model: str,
+    prompt: str,
+    api_url: str,
+    prompt_len: int,
+    output_len: int,
+    best_of: int,
+    use_beam_search: bool,
+    **kwargs,
+):
+    output = {}
+    output["prompt_len"] = prompt_len
+    oai_client = AsyncOpenAI(
+        base_url=api_url, api_key=os.environ.get("OPENAI_API_KEY")
+    )
+
+    assert not use_beam_search
+
+    try:
+        st = time.perf_counter()
+        resp = await oai_client.completions.create(
+            model=model,
+            prompt=prompt,
+            temperature=0,
+            max_tokens=output_len,
+            best_of=best_of,
+        )
+        latency = time.perf_counter() - st
+        output["generated_text"] = resp.choices[0].text
+        output["success"] = True
+        output["latency"] = latency
+    except:
+        output["generated_text"] = ""
+        output["success"] = False
+
+    return output
+
+
+ASYNC_REQUEST_FUNCS = {
```

Comment:
nit: Do we want to organize these into a class? It may make this a lot cleaner and we can then generate an interface for all future backend benchmarks, which will keep the main class free of major changes if new backends are added.

Reply:
My thought is to keep this particular file as flexible as possible: If someone wants to add support for a new backend "ABC", the only thing they need to do is to add an async function for "ABC" that performs an online inference on a given prompt (could be http or grpc) then add `"ABC"` to `ASYNC_REQUEST_FUNCS` without needing to touch the main benchmark script.

We could refactor these async request functions into a class, but again that'll still require implementations of class methods to send request & parse outputs, which are essentially what these functions are doing.